### DISCUSSION TITLE
Use AzureNotificationHubs without the client creating the registration/installation

### DISCUSSION THREAD

**Discussion by bryansjenks** [shinyorg/shiny#1293] [Q&A] [answered]
Is there a way to use Shiny for remote push notifications without the app creating it's own installation/registration in Azure notification hubs?
In order to really use the features of Azure Notification Hubs, you have to update the "Installation" with templates which is easier to do from the back-end. At that point, the mobile application doesn't need to talk to AzureNotificationHubs directly and the connection string and overhead of the API call could be removed entirely - if there was a way to opt-out of that step.

**Comment by aritchie** [comment, marked as answer]
Just use Shiny.Push for the native registration stuff and work with azure notification hubs however you want.

**Reply by bryansjenks** [reply]
This worked, thank you very much!

I will note for anyone else attempting it, you do still have to include `google-services.json` in the project and I had an issue where I couldn't set the build action on the `google-services.json` file to `GoogleServicesJson` (because it wasn't in the list) which is necessary for the notifications to be initialized properly. I initially I thought it was because I didn't have the `Shiny.Push.AzureNotificationHubs` package added, but it might actually have been because I'm targeting `net7.0-android33.0` instead of `net7.0-android`.

In any case, I fixed it by manually editing the `.csproj` file to set the build action as shown below:

```
<ItemGroup>
	  <None Remove="google-services.json" />
          ...
</ItemGroup>

<ItemGroup>
	  <GoogleServicesJson Include="google-services.json" />
</ItemGroup>
```

**Reply by aritchie** [reply]
If you are using a separate project (not a MAUI project), you have to install Xamarin.Firebase.Messaging directly which contains the necessary google libraries that contain the msbuild tasks for GoogleServicesJson to work.  

Shiny also allows you to manual set the necessary registration variables as shown here: https://github.com/shinyorg/shiny/blob/master/samples/Sample.Push.Maui/MauiProgram.cs#L35

**Reply by aritchie** [reply]
This is now documented on how to create a custom push provider that helps with registering beyond native implementations.

That documentation exists here: https://shinylib.net/client/push/providers/

**Reply by bryansjenks** [reply]
@aritchie Thanks for being so proactive with support and for all the great work on Shiny!
